### DISCUSSION TITLE
How to convert & validate arguments?

### DISCUSSION THREAD

**Discussion by cykl** [making/yavi#286] [Q&A]
Hi,

I'm evaluating YAVI in the context of query parameter validation in a JAX-RS application. I'm struggling a bit a find a good design for this use case and some guidance would be greatly appreciated.

The problem can be reduced to this scenario:  

> A validated PaginationParams object must be constructed from a  `MultivaluedMap<String, String>` (originating from `uriInfo.getQueryParameters( )`)  

```java
public record PaginationParam(@Nullable String cursor, int limit) {
```

Validating the record itself is easy (ex: check than limit is positive and smaller than a given number). However, I wonder how to handle the String -> int conversion part. I want to report a ConstraintViolation if `limit` parameter is not a number, and set it to a default value if not set as query param. 

My understanding is that ArgumentValidator could be used for this task but I don't see and elegant way to perform the string validation ->  string to int conversion- > int validation sequence. I currently have something like this that I don't really like:

```java
public record PaginationParam(
        @Nullable String cursor,
        int maxPageSize) {

    public static final Validator<PaginationParam> VALIDATOR = ValidatorBuilder.<PaginationParam> of( )
            .constraint( PaginationParam::maxPageSize, "limit", c -> c.positive( ).lessThanOrEqual( 1_000 ) )
            .constraint( PaginationParam::cursor, "cursor", c -> c.notBlank( ) )
            .build( );

    public static PaginationParam validatedFromQueryParams( MultivaluedMap<String, String> queryParams ) {
        var limitValidator = StringValidatorBuilder
                .of( "limit", c -> c.isInteger( ) )
                .build( )
                .andThen( s -> s != null ? Integer.parseInt( s ) : 1_000 );

        var cursorValidator = StringValidatorBuilder
                .of( "cursor", c -> c )
                .build( );

        var paramsValidator = ArgumentsValidators.split( cursorValidator, limitValidator )
                .apply( PaginationParam::new );

        var paginationParams = paramsValidator.validate(
                queryParams.getFirst( "cursor" ),
                queryParams.getFirst( "limit" ) )
                .orElseThrow( PaginationParam::asProblem );

        return VALIDATOR.applicative( ).validate( paginationParams )
                .orElseThrow( PaginationParam::asProblem );
    }

    public static ConstraintViolationProblem asProblem( List<ConstraintViolation> violations ) {
        return new ConstraintViolationProblem( BAD_REQUEST, violations.stream( )
                .map( violation -> new Violation( violation.name( ), violation.message( ) ) )
                .toList( ) );
    }
```

Issues:
* All violations should be reported rather than having two steps
* Field name is duplicated between the Validator<PaginationParam>  and ArgumentValidator

Is there a better way to do it? Can I move  the `.constraint( PaginationParam::maxPageSize, "limit", c -> c.positive( ).lessThanOrEqual( 1_000 ) )` after `andThen` and ideally avoid field name duplication?

My challenge is that limit is an integer

**Comment by making** [comment]
@cykl I'm sorry that I wasn't aware of this thread.

I would write as follows (using YAVI 0.14)

```java
Arguments1Validator<String, Integer> limitValidatgor = StringValidatorBuilder.of("limit", c -> c.isInteger())
		.build(s -> s != null && !s.isBlank() ? Integer.parseInt(s) : 1_000)
		.andThen(IntegerValidatorBuilder.of("limit", c -> c.notNull().positive().lessThanOrEqual(1_000)).build());

Arguments1Validator<MultiValueMap<String, String>, PaginationParam> validator = Yavi.arguments()
		._string("cursor", c -> c.notBlank())
		._string(limitValidatgor)
		.apply(PaginationParam::new)
		.compose(queryParams -> Arguments.of(queryParams.getFirst("cursor"), queryParams.getFirst("limit")));
```
